java-api-core/src/com/kinvey/java/store/RealtimeRouter.java

```diff
@@ -0,0 +1,99 @@
+package com.kinvey.java.store;
+
+import com.kinvey.java.AbstractClient;
+import com.pubnub.api.PNConfiguration;
+import com.pubnub.api.PubNub;
+import com.pubnub.api.callbacks.SubscribeCallback;
+import com.pubnub.api.models.consumer.PNStatus;
+import com.pubnub.api.models.consumer.pubsub.PNMessageResult;
+import com.pubnub.api.models.consumer.pubsub.PNPresenceEventResult;
+
+import java.util.Collections;
+
+/**
+ * Created by yuliya on 2/15/17.
+ */
+
+class RealtimeRouter {
+
+    private static final Object lock = new Object();
+    private static RealtimeRouter realtimeRouter;
+    private PubNub pubnubClient;
+    private String channelGroup;
+    private AbstractClient client;
+    private SubscribeCallback subscribeCallback;
+
+    private RealtimeRouter() {
+
+    }
+
+    static RealtimeRouter getInstance() {
+        synchronized (lock) {
+            if (realtimeRouter == null) {
+                realtimeRouter = new RealtimeRouter();
+            }
+            return realtimeRouter;
+        }
+    }
+
+    void initialize(String channelGroup, String publishKey, String subscribeKey, String authKey, AbstractClient client) {
+        synchronized (lock) {
+            if (!isInitialized()) {
+                this.channelGroup = channelGroup;
+                this.client = client;
+                PNConfiguration pnConfiguration = new PNConfiguration();
+                pnConfiguration.setSubscribeKey(subscribeKey);
+                pnConfiguration.setPublishKey(publishKey);
+                pnConfiguration.setAuthKey(authKey);
+                pnConfiguration.setSecure(true);
+                pubnubClient = new PubNub(pnConfiguration);
+                subscribeCallback = new SubscribeCallback() {
+                    @Override
+                    public void status(PubNub pubnub, PNStatus status) {
+                        handleStatusMessage(status);
+                    }
+
+                    @Override
+                    public void message(PubNub pubnub, PNMessageResult message) {
+                        subscribeCallback(message.getChannel(), message.getMessage().getAsString());
+                    }
+
+                    @Override
+                    public void presence(PubNub pubnub, PNPresenceEventResult presence) {
+                    /* presence not currently supported */
+                    }
+                };
+                pubnubClient.addListener(subscribeCallback);
+                pubnubClient.subscribe().channelGroups(Collections.singletonList(channelGroup)).execute();
+            }
+        }
+
+    }
+
+    void unInitialize() {
+        synchronized (lock) {
+            if (isInitialized()) {
```

Comment:
Same question here about locking as in the `initialize()` method.

Reply:
Updated.